### PR TITLE
Add QuartzNET instrumentation

### PR DESCRIPTION
**Pull request by maldago**
Hello,

Here is an instrumentation implementation for [Quartz.NET](https://github.com/quartznet/quartznet)

Quartz.NET had originally done it's own instrumentation implementation based on OpenTelemetry 0.6.0-beta.1, which could not longer be maintained as the DiagnosticSource helpers were internalized. This is intended to replace the existing Quartz.NET implementation and that can be removed from the Quartz.NET source.

Thanks

### CODE REVIEW

File: src/OpenTelemetry.Contrib.Instrumentation.Quartz/Implementation/QuartzDiagnosticListener.cs
Comment (cijothomas) [inline src/OpenTelemetry.Contrib.Instrumentation.Quartz/Implementation/QuartzDiagnosticListener.cs:48]: you can skip this check completely. Its already handled by the SDK.

File: src/OpenTelemetry.Contrib.Instrumentation.Quartz/README.md
Comment (cijothomas) [inline src/OpenTelemetry.Contrib.Instrumentation.Quartz/README.md:11]: it'd be good to include usage example as well here.

File: src/OpenTelemetry.Contrib.Instrumentation.Quartz/QuartzInstrumentationOptions.cs
Comment (cijothomas) [inline src/OpenTelemetry.Contrib.Instrumentation.Quartz/QuartzInstrumentationOptions.cs:29]: does this need to be public?

File: src/OpenTelemetry.Contrib.Instrumentation.Quartz/QuartzInstrumentationOptions.cs
Comment (cijothomas) [inline src/OpenTelemetry.Contrib.Instrumentation.Quartz/QuartzInstrumentationOptions.cs:36]:
```suggestion
        /// Gets or sets a value indicating whether gets or sets a value whether to add exception details as ActivityEvents.
```

File: src/OpenTelemetry.Contrib.Instrumentation.Quartz/Implementation/QuartzDiagnosticListener.cs
Comment (cijothomas) [inline src/OpenTelemetry.Contrib.Instrumentation.Quartz/Implementation/QuartzDiagnosticListener.cs:75]: nothing seem to be done here :)

File: src/OpenTelemetry.Contrib.Instrumentation.Quartz/Implementation/QuartzDiagnosticListener.cs
Comment (cijothomas) [inline src/OpenTelemetry.Contrib.Instrumentation.Quartz/Implementation/QuartzDiagnosticListener.cs:97]: https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Api/Trace/ActivityExtensions.cs#L77 For exception reporting, please use this method.

Comment (cijothomas) [reply src/OpenTelemetry.Contrib.Instrumentation.Quartz/Implementation/QuartzDiagnosticListener.cs:97]: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/exceptions.md#attributes

File: src/OpenTelemetry.Contrib.Instrumentation.Quartz/OpenTelemetry.Contrib.Instrumentation.Quartz.csproj
Comment (cijothomas) [inline src/OpenTelemetry.Contrib.Instrumentation.Quartz/OpenTelemetry.Contrib.Instrumentation.Quartz.csproj:11]: you should be able to take just a API dependency instead of SDK, once the other comments are addressed.

**Comment by cijothomas** [comment]
@maldago THanks for your contribution! It looks like the EASYCla is failing.. Can you try to fix it?

File: src/OpenTelemetry.Contrib.Instrumentation.Quartz/README.md
Comment (maldago) [reply src/OpenTelemetry.Contrib.Instrumentation.Quartz/README.md:11]: udpated

File: src/OpenTelemetry.Contrib.Instrumentation.Quartz/QuartzInstrumentationOptions.cs
Comment (maldago) [reply src/OpenTelemetry.Contrib.Instrumentation.Quartz/QuartzInstrumentationOptions.cs]: nope, not at this time

File: src/OpenTelemetry.Contrib.Instrumentation.Quartz/Implementation/QuartzDiagnosticListener.cs
Comment (maldago) [reply src/OpenTelemetry.Contrib.Instrumentation.Quartz/Implementation/QuartzDiagnosticListener.cs]: very true!

File: src/OpenTelemetry.Contrib.Instrumentation.Quartz/Implementation/QuartzDiagnosticListener.cs
Comment (maldago) [reply src/OpenTelemetry.Contrib.Instrumentation.Quartz/Implementation/QuartzDiagnosticListener.cs]: updated to use activity extension

File: src/OpenTelemetry.Contrib.Instrumentation.Quartz/OpenTelemetry.Contrib.Instrumentation.Quartz.csproj
Comment (maldago) [reply src/OpenTelemetry.Contrib.Instrumentation.Quartz/OpenTelemetry.Contrib.Instrumentation.Quartz.csproj]: updated from SDK to API

File: .github/workflows/package-Instrumentation.Quartz.yml
Comment (cijothomas) [inline .github/workflows/package-Instrumentation.Quartz.yml:49]: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/131/files  -- please pick this too!

File: src/OpenTelemetry.Contrib.Instrumentation.Quartz/Implementation/QuartzInstrumentationEventSource.cs
Comment (cijothomas) [inline src/OpenTelemetry.Contrib.Instrumentation.Quartz/Implementation/QuartzInstrumentationEventSource.cs:53]: not used anywhere?

File: src/OpenTelemetry.Contrib.Instrumentation.Quartz/Implementation/QuartzInstrumentationEventSource.cs
Comment (cijothomas) [inline src/OpenTelemetry.Contrib.Instrumentation.Quartz/Implementation/QuartzInstrumentationEventSource.cs:38]: nit: rename to more apt name for this library.

File: src/OpenTelemetry.Contrib.Instrumentation.Quartz/Implementation/QuartzDiagnosticListener.cs
Comment (cijothomas) [inline src/OpenTelemetry.Contrib.Instrumentation.Quartz/Implementation/QuartzDiagnosticListener.cs:67]: this seems different that the fetcher approach used in other instrumentations.. https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs#L271

Comment (cijothomas) [reply src/OpenTelemetry.Contrib.Instrumentation.Quartz/Implementation/QuartzDiagnosticListener.cs:67]: i'd recommend a unit test to see if this works.

Comment (maldago) [reply src/OpenTelemetry.Contrib.Instrumentation.Quartz/Implementation/QuartzDiagnosticListener.cs:67]: @cijothomas , I have been looking at the example you have linked. 

`  private readonly PropertyFetcher<Exception> stopExceptionFetcher = new PropertyFetcher<Exception>("Exception");`

When processing the `stopExceptionFetcher.TryFetch`
`  if (!this.stopExceptionFetcher.TryFetch(payload, out Exception exception) || exception == null) `

The fetch returns null even though the `payload` is of type `Exception`. I think that this is because the `stopExceptionFetcher` is looking for a property `Exception` on the exception object.

File: .github/workflows/package-Instrumentation.Quartz.yml
Comment (maldago) [reply .github/workflows/package-Instrumentation.Quartz.yml]: picked|!

File: src/OpenTelemetry.Contrib.Instrumentation.Quartz/Implementation/QuartzInstrumentationEventSource.cs
Comment (maldago) [reply src/OpenTelemetry.Contrib.Instrumentation.Quartz/Implementation/QuartzInstrumentationEventSource.cs:53]: it is now used, updated instrumentation

File: src/OpenTelemetry.Contrib.Instrumentation.Quartz/Implementation/QuartzInstrumentationEventSource.cs
Comment (maldago) [reply src/OpenTelemetry.Contrib.Instrumentation.Quartz/Implementation/QuartzInstrumentationEventSource.cs]: renamed to give more context

File: src/OpenTelemetry.Contrib.Instrumentation.Quartz/TraceProviderBuilderExtensions.cs
Comment (cijothomas) [inline src/OpenTelemetry.Contrib.Instrumentation.Quartz/TraceProviderBuilderExtensions.cs:43]: does Quartz library publish DiagnosticSource events? Could you point to the source code where it is doing so? I'd assume its already creating Activity, using legacy way, and hence we need to run them through samplers, etc, by using AddLegacySource API..
https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/docs/trace/extending-the-sdk#special-case--instrumentation-for-libraries-producing-legacy-activity

Comment (maldago) [reply src/OpenTelemetry.Contrib.Instrumentation.Quartz/TraceProviderBuilderExtensions.cs:43]: Quartz  library publishes DiagnosticSource events:
https://github.com/quartznet/quartznet/blob/main/src/Quartz/Logging/JobDiagnosticsWriter.cs

Comment (cijothomas) [reply src/OpenTelemetry.Contrib.Instrumentation.Quartz/TraceProviderBuilderExtensions.cs:43]: Okay so it does start the Activity. I doubt if the activity actually goes through the samplers, as you are not calling `AddLegacySource` for this activity.

Comment (cijothomas) [reply src/OpenTelemetry.Contrib.Instrumentation.Quartz/TraceProviderBuilderExtensions.cs:43]: cc: @utpilla

Comment (lahma) [reply src/OpenTelemetry.Contrib.Instrumentation.Quartz/TraceProviderBuilderExtensions.cs:43]: On Quartz side we can also change things per recommendations how the current state of things should be. The current beta-based implementation has been marketed as unstable API etc so happy to steer towards ideal solution.

Comment (cijothomas) [reply src/OpenTelemetry.Contrib.Instrumentation.Quartz/TraceProviderBuilderExtensions.cs:43]: if that is a possibility,  I strongly recommend that :) 
https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/docs/trace/extending-the-sdk#writing-own-instrumentation-library

It'd involve using ActivitySource model of creating activities. Once the Quartz library does that, then *this* instrumentation is optional. :)

Comment (lahma) [reply src/OpenTelemetry.Contrib.Instrumentation.Quartz/TraceProviderBuilderExtensions.cs:43]: I'll let @maldago steer this, I don't have enough subject expertise and grateful for him doing the effort 👍🏻

**Comment by maldago** [comment]
Hello,

What is the status of this PR? Is there more work expected from my end? Looking for an update

Cheers

File: src/OpenTelemetry.Contrib.Instrumentation.Quartz/OpenTelemetry.Contrib.Instrumentation.Quartz.csproj
Comment (cijothomas) [inline src/OpenTelemetry.Contrib.Instrumentation.Quartz/OpenTelemetry.Contrib.Instrumentation.Quartz.csproj:10]: can you pick 1.1.0 as 1.1.0 stable has released!

File: src/OpenTelemetry.Contrib.Instrumentation.Quartz/README.md
Comment (cijothomas) [inline src/OpenTelemetry.Contrib.Instrumentation.Quartz/README.md:34]: this requires using ext.hosting package.. could you update the readme to indicate that?

File: src/OpenTelemetry.Contrib.Instrumentation.Quartz/OpenTelemetry.Contrib.Instrumentation.Quartz.csproj
Comment (cijothomas) [inline src/OpenTelemetry.Contrib.Instrumentation.Quartz/OpenTelemetry.Contrib.Instrumentation.Quartz.csproj:7]: any reason why you want to separately target net461 and net472?

File: test/OpenTelemetry.Contrib.Instrumentation.Quartz.Tests/OpenTelemetry.Contrib.Instrumentation.Quartz.Tests.csproj
Comment (cijothomas) [inline test/OpenTelemetry.Contrib.Instrumentation.Quartz.Tests/OpenTelemetry.Contrib.Instrumentation.Quartz.Tests.csproj:5]: we can remove netcoreapp2.1, as its EOL very soon! (the main repo is in the process of removing it already).

Also, do you want to add atleast one .NET Framework target for the tests/

File: src/OpenTelemetry.Contrib.Instrumentation.Quartz/README.md
Comment (cijothomas) [inline src/OpenTelemetry.Contrib.Instrumentation.Quartz/README.md:43]: could you add Enrich also to the readme?

File: test/OpenTelemetry.Contrib.Instrumentation.Quartz.Tests/QuartzDiagnosticListenerTests.cs
Comment (cijothomas) [inline test/OpenTelemetry.Contrib.Instrumentation.Quartz.Tests/QuartzDiagnosticListenerTests.cs:210]: when RecordException is true, please validate that the exception are stored as activityevent. (the activity.RecordException does just that, so would be good to validate that in unit tests)

File: test/OpenTelemetry.Contrib.Instrumentation.Quartz.Tests/QuartzDiagnosticListenerTests.cs
Comment (cijothomas) [inline test/OpenTelemetry.Contrib.Instrumentation.Quartz.Tests/QuartzDiagnosticListenerTests.cs:108]: We should be doing this fetching ourselves, and not require users to do it.

File: src/OpenTelemetry.Contrib.Instrumentation.Quartz/Implementation/QuartzDiagnosticListener.cs
Comment (cijothomas) [inline src/OpenTelemetry.Contrib.Instrumentation.Quartz/Implementation/QuartzDiagnosticListener.cs:60]: similar instrumentations we have (http, aspnetcore), extracts the required objects from the payload using the property fetchers, so end user always get the right type, and don't have to deal with the payload extraction themselves. Any reason why we don't that approach here? 
Same comment for exception as well..

Comment (cijothomas) [reply src/OpenTelemetry.Contrib.Instrumentation.Quartz/Implementation/QuartzDiagnosticListener.cs:60]: @CodeBlanch to share your thoughts on this approach, where the payload is passed as-is to the Enrich, vs extracting the strong type object (JobDetail etc), and passing it to Enrich.

**Comment by cijothomas** [comment]
> Hello,
> 
> What is the status of this PR? Is there more work expected from my end? Looking for an update
> 
> Cheers

I just did another round of review on this PR. Could you check my comments?

(Sorry for delay here, most of the energy is spent in the Metrics work, so please expect some delays. Please let us know when you get a chance to respond to the comments shared above.)

File: src/OpenTelemetry.Contrib.Instrumentation.Quartz/OpenTelemetry.Contrib.Instrumentation.Quartz.csproj
Comment (maldago) [reply src/OpenTelemetry.Contrib.Instrumentation.Quartz/OpenTelemetry.Contrib.Instrumentation.Quartz.csproj]: updated to 1.1.0 stable

File: src/OpenTelemetry.Contrib.Instrumentation.Quartz/README.md
Comment (maldago) [reply src/OpenTelemetry.Contrib.Instrumentation.Quartz/README.md]: updated usage & README

File: test/OpenTelemetry.Contrib.Instrumentation.Quartz.Tests/OpenTelemetry.Contrib.Instrumentation.Quartz.Tests.csproj
Comment (maldago) [reply test/OpenTelemetry.Contrib.Instrumentation.Quartz.Tests/OpenTelemetry.Contrib.Instrumentation.Quartz.Tests.csproj]: removed netcoreapp2.1 and added net472

File: test/OpenTelemetry.Contrib.Instrumentation.Quartz.Tests/QuartzDiagnosticListenerTests.cs
Comment (maldago) [reply test/OpenTelemetry.Contrib.Instrumentation.Quartz.Tests/QuartzDiagnosticListenerTests.cs:211]: added assertion for `ActivityEvent.Name`  name

File: src/OpenTelemetry.Contrib.Instrumentation.Quartz/OpenTelemetry.Contrib.Instrumentation.Quartz.csproj
Comment (maldago) [reply src/OpenTelemetry.Contrib.Instrumentation.Quartz/OpenTelemetry.Contrib.Instrumentation.Quartz.csproj]: cleaned up

File: test/OpenTelemetry.Contrib.Instrumentation.Quartz.Tests/QuartzDiagnosticListenerTests.cs
Comment (maldago) [reply test/OpenTelemetry.Contrib.Instrumentation.Quartz.Tests/QuartzDiagnosticListenerTests.cs]: fetched internally, updated  README to reflect changes to be inline with other implementations

File: src/OpenTelemetry.Contrib.Instrumentation.Quartz/TraceProviderBuilderExtensions.cs
Comment (cijothomas) [inline src/OpenTelemetry.Contrib.Instrumentation.Quartz/TraceProviderBuilderExtensions.cs:45]: Is this list static, or does it have to account for `TracedOperations` ?

File: src/OpenTelemetry.Contrib.Instrumentation.Quartz/OpenTelemetry.Contrib.Instrumentation.Quartz.csproj
Comment (cijothomas) [inline src/OpenTelemetry.Contrib.Instrumentation.Quartz/OpenTelemetry.Contrib.Instrumentation.Quartz.csproj:7]: if Quartz supports, then consider targeting net461 as that's an officially supported .NET version.

File: src/OpenTelemetry.Contrib.Instrumentation.Quartz/README.md
Comment (cijothomas) [inline src/OpenTelemetry.Contrib.Instrumentation.Quartz/README.md:55]: would be nice to provide a list of supported, or is this free form and can be anything?

File: src/OpenTelemetry.Contrib.Instrumentation.Quartz/README.md
Comment (cijothomas) [inline src/OpenTelemetry.Contrib.Instrumentation.Quartz/README.md:82]: `IJobDetail` require a different namespace, would be good to add it.

File: src/OpenTelemetry.Contrib.Instrumentation.Quartz/OpenTelemetry.Contrib.Instrumentation.Quartz.csproj
Comment (cijothomas) [inline src/OpenTelemetry.Contrib.Instrumentation.Quartz/OpenTelemetry.Contrib.Instrumentation.Quartz.csproj:3]: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/CODEOWNERS Could you update this to list you as an owner.

**Comment by cijothomas** [comment]
I have left a few minor comments. We can merge this, and address them separately. Would like to have someone as owner for this contribution.

File: src/OpenTelemetry.Contrib.Instrumentation.Quartz/OpenTelemetry.Contrib.Instrumentation.Quartz.csproj
Comment (maldago) [reply src/OpenTelemetry.Contrib.Instrumentation.Quartz/OpenTelemetry.Contrib.Instrumentation.Quartz.csproj]: this has been updated

File: src/OpenTelemetry.Contrib.Instrumentation.Quartz/README.md
Comment (maldago) [reply src/OpenTelemetry.Contrib.Instrumentation.Quartz/README.md]: updated to succinct supported list

File: src/OpenTelemetry.Contrib.Instrumentation.Quartz/OpenTelemetry.Contrib.Instrumentation.Quartz.csproj
Comment (maldago) [reply src/OpenTelemetry.Contrib.Instrumentation.Quartz/OpenTelemetry.Contrib.Instrumentation.Quartz.csproj:3]: updated

File: src/OpenTelemetry.Contrib.Instrumentation.Quartz/README.md
Comment (maldago) [reply src/OpenTelemetry.Contrib.Instrumentation.Quartz/README.md]: add namespace

**Comment by maldago** [comment]
> I have left a few minor comments. We can merge this, and address them separately. Would like to have someone as owner for this contribution.

I have finally had the time to update the PR with the relevant changes. 

Thanks

File: src/OpenTelemetry.Contrib.Instrumentation.Quartz/README.md
Comment (cijothomas) [inline src/OpenTelemetry.Contrib.Instrumentation.Quartz/README.md:38]: nit: pre-requesite is not mentioned to add this Jaeger package.

File: src/OpenTelemetry.Contrib.Instrumentation.Quartz/README.md
Comment (cijothomas) [inline src/OpenTelemetry.Contrib.Instrumentation.Quartz/README.md:46]:
```suggestion
This option allows you to filter trace operations.

```

**Review by cijothomas — APPROVED** [review]

**Comment by cijothomas** [comment]
> > I have left a few minor comments. We can merge this, and address them separately. Would like to have someone as owner for this contribution.
> 
> I have finally had the time to update the PR with the relevant changes.
> 
> Thanks

LGTM. We can merge once the CI is green. Please check the failures.
Also, please do a follow up to rename the project/namespaces to remove "Contrib" from it, similar to https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/213, to align with https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/196

**Comment by cijothomas** [comment]
@maldago Could you check CI failures?

**Comment by maldago** [comment]
> @maldago Could you check CI failures?

I have done so. I had to update the `net461` target to `net472`, as Quartz does not support activity sources in `net461`, as details [here](https://www.quartz-scheduler.net/2021/04/07/quartznet-3-3-released/). The README.md has been updated to reflect which frameworks are supported.  I updated the proj/namespaces.

**Comment by maldago** [comment]
> @maldago Retriggered CIs. Please check if there are reds.

there was a couple of small formatting issues identified by sanity check, should be all good now

**Comment by cijothomas** [comment]
@maldago Merged :) Please check the below comment as well

Also, please do a follow up to rename the project/namespaces to remove "Contrib" from it, similar to https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/213, to align with https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/196

**Comment by maldago** [comment]
> @maldago Merged :) Please check the below comment as well
> 
> Also, please do a follow up to rename the project/namespaces to remove "Contrib" from it, similar to #213, to align with #196

Already in the PR you merged :)

**Comment by cijothomas** [comment]
> > @maldago Merged :) Please check the below comment as well
> > Also, please do a follow up to rename the project/namespaces to remove "Contrib" from it, similar to #213, to align with #196
> 
> Already in the PR you merged :)

Thanks for confirmation. Will do the 1st release to nuget this week.

**Pull request merged**
